### DISCUSSION TITLE
Diagonal Selectors

### DISCUSSION THREAD

**Discussion by kiryph** [lvjr/tabularray#288] [Ideas]
I am creating an adjacency matrix of a graph and I want to highlight the diagonal line (each node is a neighbour of itself):
![Z2^3;_Lattice_of_subgroups_Hasse_diagram_adjacency_matrix svg](https://user-images.githubusercontent.com/371756/184631601-0d984bcd-7e65-41aa-9b65-516f090d024d.png)

https://commons.wikimedia.org/wiki/File:Z2%5E3;_Lattice_of_subgroups_Hasse_diagram_adjacency_matrix.svg

Could we have a convenient selector for the diagonal?

I am creating matrices with up to 100 rows and columns which makes it very verbose when the number of single cells which should be marked with a light red color is 100.

**Comment by lvjr** [comment]
Maybe it is better to use `pgf/tikz` to draw this graph. It is a little slow to build a 100 by 100 table with `tabularray`. But at least it is possible to highlight the diagonal cells with `functional` library:
```latex
\documentclass{article}
\usepackage{tabularray}
\usepackage{xcolor}
\UseTblrLibrary{functional}
\IgnoreSpacesOn
\prgNewFunction \myStyling {} {
  \intStepOneInline {1} {\arabic{rowcount}} {
    \cellSetStyle {##1} {##1} {bg=azure8}
  }
}
\IgnoreSpacesOff
\begin{document}
\begin{tblr}{process=\myStyling}
  Alpha   & Beta  & Gamma   & Delta   \\
  Epsilon & Zeta  & Eta     & Theta   \\
  Iota    & Kappa & Lambda  & Mu      \\
  Nu      & Xi    & Omicron & Pi      \\
\end{tblr}
\end{document}
```
![image](https://user-images.githubusercontent.com/4239949/185020443-39650388-8ab3-40b3-bd35-2b6d6ccd6109.png)

See the manual of [`tabularray`](http://mirrors.ctan.org/macros/latex/contrib/tabularray/tabularray.pdf) and [`functional`](http://mirrors.ctan.org/macros/latex/contrib/functional/functional.pdf) packages for more details.

**Reply by kiryph** [reply]
Thanks for the answer and pointing out that pgf/tikz might be a better choice. This was one of the questions I actually also had. Do you have tikz-matrix in mind or a manual approach?

**Reply by lvjr** [reply]
Sorry, I haven't used much of `pgf`/`tikz`.

**Reply by dbitouze** [reply]
> Thanks for the answer and pointing out that pgf/tikz might be a better choice. This was one of the questions I actually also had. Do you have tikz-matrix in mind or a manual approach?

You might be interested in the [`nicematrix`](https://www.ctan.org/pkg/nicematrix) package.

**Reply by kiryph** [reply]
@dbitouze Thanks for your suggestion. Is there something specific why you recommend `nicematrix`?

For instance, a particular feature in my context, the performance, ease of use, a stable, well-maintained and -received package, personal experience, ... I am honestly interested.

I have started reading the documentation and it looks like it that it has it all (i.e. very feature rich):

- cell background colouring
- rotation of the text in the header row with `\rotate`
- top left cell can be divided with `\diagbox`
- ...

I had issues with the expansion of macros in `tabularray` which are solved via several means

- expand a single macro as following
```tex
\begin{blr}[expand=\rows]{colspec={|c|c|c|c|}}
\rows
\end{tblr}
```
- use the package [functional](https://ctan.org/pkg/functional) for
  - macros in the mandatory argument (see above) and
  - within rows see [tex.stackexchange answer](https://tex.stackexchange.com/a/645116/8917)
-  [rewrite environment](https://tex.stackexchange.com/a/654234/8917)  to expand macros in the mandatory argument

However, note there is a pending issue about [expanding multiple macros](https://github.com/lvjr/tabularray/issues/77).

Performance-wise I am not sure whether `nicematrix` is faster since it needs two compilation runs. But it all depends how their code handle larger tables with more than 50 rows and columns. My first table is 27x27 but it will not be the last one and others will be larger. So I would prefer a solution which can handle it better.

However, I have started now with `tabularray` and I will finish the current one and will see how the next larger one goes with `tabularray`. If compilation time will become unbaringly slow, I will rewrite with `nicematrix`.

**Reply by kiryph** [reply]
Here some timings
- 27x27 table took 10 seconds to create
- 40x40 table took 21 seconds to create

  ![Screenshot 2022-08-18 at 11 06 27](https://user-images.githubusercontent.com/371756/185357630-cea41914-4642-43d5-a45e-7fc799f3e33d.png)


- 112x112 table took three minutes and 36 seconds. Here it becomes so slow that one can consider investing time in a faster solution. However, these are so large that I am not sure whether I will actually use them:
  ![Screenshot 2022-08-18 at 11 08 22](https://user-images.githubusercontent.com/371756/185357666-fbd70008-dcbe-44c4-9beb-c6cad1cd88b7.png)


  Most importantly, I intend to remove columns with only trivial subgroups (group itself). This removes a large area with no information.

**Reply by dbitouze** [reply]
@kiryph 

> Is there something specific why you recommend `nicematrix`?

Just because:

1. you want to typeset matrices,
2. you said “Thanks for the answer and pointing out that `pgf`/`tikz` might be a better choice.”,
3. I'm aware `nicematrix` typesets matrices relying on `pgf`/`tikz`.

:smile:
